src/Domain/Neighbors.cpp

```diff
@@ -47,19 +44,30 @@ void Neighbors<VolumeDim>::pup(PUP::er& p) noexcept {
   p | orientation_;
 }
 
+template <size_t VolumeDim>
+std::ostream& operator<<(std::ostream& os,
+                         const domain::Neighbors<VolumeDim>& n) {
+  ::operator<<(os << "Ids = ", n.ids())
+      << "; orientation = " << n.orientation();
+  return os;
+}
+}  // namespace domain
 // Explicit instantiations
-template class Neighbors<1>;
-template class Neighbors<2>;
-template class Neighbors<3>;
 
-template std::ostream& operator<<(std::ostream&, const Neighbors<1>&);
-template std::ostream& operator<<(std::ostream&, const Neighbors<2>&);
-template std::ostream& operator<<(std::ostream&, const Neighbors<3>&);
+#define DIM(data) BOOST_PP_TUPLE_ELEM(0, data)
+
+#define INSTANTIATE(_, data)                                        \
+  template class domain::Neighbors<DIM(data)>;                      \
+  template std::ostream& domain::operator<<(                        \
+      std::ostream& os, const domain::Neighbors<DIM(data)>& block); \
+  template bool domain::operator==(                                 \
+      const domain::Neighbors<DIM(data)>& lhs,                      \
+      const domain::Neighbors<DIM(data)>& rhs) noexcept;            \
+  template bool domain::operator!=(                                 \
+      const domain::Neighbors<DIM(data)>& lhs,                      \
+      const domain::Neighbors<DIM(data)>& rhs) noexcept;
 
-template bool operator==(const Neighbors<1>&, const Neighbors<1>&) noexcept;
-template bool operator==(const Neighbors<2>&, const Neighbors<2>&) noexcept;
-template bool operator==(const Neighbors<3>&, const Neighbors<3>&) noexcept;
+GENERATE_INSTANTIATIONS(INSTANTIATE, (1, 2, 3))
```

Comment:
Factor out.  (And can't you instantiate inside the namespace and avoid all the qualifiers?)

Reply:
No, and yes